File: concepts/02-linked-lists/node/linkedList.js
/**
 * Linked List Implementation in Node.js
 */

class ListNode {
    /**
     * Node class for singly linked list
     */
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

class SinglyLinkedList {
    /**
     * Singly Linked List Implementation
     */
    constructor() {
        this.head = null;
        this.size = 0;
    }

    /**
     * Add node at the end - O(n)
     */
    append(val) {
        const newNode = new ListNode(val);
        if (!this.head) {
            this.head = newNode;
        } else {
            let current = this.head;
            while (current.next) {
                current = current.next;
            }
            current.next = newNode;
        }
        this.size++;
    }

    /**
     * Add node at the beginning - O(1)
     */
    prepend(val) {
        const newNode = new ListNode(val);
        newNode.next = this.head;
        this.head = newNode;
        this.size++;
    }

    /**
     * Insert node at index - O(n)
     */
    insertAt(index, val) {
        if (index < 0 || index > this.size) {
            throw new Error("Index out of range");
        }

        if (index === 0) {
            this.prepend(val);
            return;
        }

        const newNode = new ListNode(val);
        let current = this.head;
        for (let i = 0; i < index - 1; i++) {
            current = current.next;
        }

        newNode.next = current.next;
        current.next = newNode;
        this.size++;
    }

    /**
     * Delete first occurrence of value - O(n)
     */
    delete(val) {
        if (!this.head) {
            return false;
        }

        if (this.head.val === val) {
            this.head = this.head.next;
            this.size--;
            return true;
        }

        let current = this.head;
        while (current.next) {
            if (current.next.val === val) {
                current.next = current.next.next;
                this.size--;
                return true;
            }
            current = current.next;
        }

        return false;
    }

    /**
     * Search for value - O(n)
     */
    search(val) {
        let current = this.head;
        let index = 0;
        while (current) {
            if (current.val === val) {
                return index;
            }
            current = current.next;
            index++;
        }
        return -1;
    }

    /**
     * Reverse linked list - O(n)
     */
    reverse() {
        let prev = null;
        let current = this.head;

        while (current) {
            const nextNode = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
        }

        this.head = prev;
    }

    /**
     * Find middle element using two pointers - O(n)
     */
    findMiddle() {
        let slow = this.head;
        let fast = this.head;

        while (fast && fast.next) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow ? slow.val : null;
    }

    /**
     * Detect cycle using Floyd's algorithm - O(n)
     */
    hasCycle() {
        let slow = this.head;
        let fast = this.head;

        while (fast && fast.next) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow === fast) {
                return true;
            }
        }

        return false;
    }

    /**
     * Display all elements
     */
    display() {
        const result = [];
        let current = this.head;
        while (current) {
            result.push(current.val);
            current = current.next;
        }
        return result;
    }
}

class DoublyListNode {
    /**
     * Node class for doubly linked list
     */
    constructor(val = 0, prev = null, next = null) {
        this.val = val;
        this.prev = prev;
        this.next = next;
    }
}

class DoublyLinkedList {
    /**
     * Doubly Linked List Implementation
     */
    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    /**
     * Add node at the end - O(1)
     */
    append(val) {
        const newNode = new DoublyListNode(val);

        if (!this.head) {
            this.head = this.tail = newNode;
        } else {
            newNode.prev = this.tail;
            this.tail.next = newNode;
            this.tail = newNode;
        }

        this.size++;
    }

    /**
     * Add node at the beginning - O(1)
     */
    prepend(val) {
        const newNode = new DoublyListNode(val);

        if (!this.head) {
            this.head = this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }

        this.size++;
    }

    /**
     * Delete first occurrence - O(n)
     */
    delete(val) {
        let current = this.head;

        while (current) {
            if (current.val === val) {
                if (current.prev) {
                    current.prev.next = current.next;
                } else {
                    this.head = current.next;
                }

                if (current.next) {
                    current.next.prev = current.prev;
                } else {
                    this.tail = current.prev;
                }

                this.size--;
                return true;
            }

            current = current.next;
        }

        return false;
    }

    /**
     * Display from head to tail
     */
    displayForward() {
        const result = [];
        let current = this.head;
        while (current) {
            result.push(current.val);
            current = current.next;
        }
        return result;
    }

    /**
     * Display from tail to head
     */
    displayBackward() {
        const result = [];
        let current = this.tail;
        while (current) {
            result.push(current.val);
            current = current.prev;
        }
        return result;
    }
}

// Example usage
if (require.main === module) {
    // Singly Linked List
    console.log("--- Singly Linked List ---");
    const sll = new SinglyLinkedList();
    sll.append(1);
    sll.append(2);
    sll.append(3);
    sll.prepend(0);
    console.log("List:", sll.display());
    console.log("Middle:", sll.findMiddle());

    sll.reverse();
    console.log("Reversed:", sll.display());

    console.log("Search 2:", sll.search(2));
    sll.delete(2);
    console.log("After delete 2:", sll.display());

    // Doubly Linked List
    console.log("\n--- Doubly Linked List ---");
    const dll = new DoublyLinkedList();
    dll.append(1);
    dll.append(2);
    dll.append(3);
    dll.prepend(0);
    console.log("Forward:", dll.displayForward());
    console.log("Backward:", dll.displayBackward());

    dll.delete(2);
    console.log("After delete 2 - Forward:", dll.displayForward());
}

module.exports = { ListNode, SinglyLinkedList, DoublyListNode, DoublyLinkedList };

File: concepts/02-linked-lists/python/linked_list.py
"""
Linked List Implementation in Python
"""

class ListNode:
    """Node class for singly linked list"""
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class SinglyLinkedList:
    """Singly Linked List Implementation"""
    
    def __init__(self):
        self.head = None
        self.size = 0
    
    def append(self, val):
        """Add node at the end - O(n)"""
        new_node = ListNode(val)
        if not self.head:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
        self.size += 1
    
    def prepend(self, val):
        """Add node at the beginning - O(1)"""
        new_node = ListNode(val)
        new_node.next = self.head
        self.head = new_node
        self.size += 1
    
    def insert_at(self, index, val):
        """Insert node at index - O(n)"""
        if index < 0 or index > self.size:
            raise IndexError("Index out of range")
        
        if index == 0:
            self.prepend(val)
            return
        
        new_node = ListNode(val)
        current = self.head
        for _ in range(index - 1):
            current = current.next
        
        new_node.next = current.next
        current.next = new_node
        self.size += 1
    
    def delete(self, val):
        """Delete first occurrence of value - O(n)"""
        if not self.head:
            return False
        
        if self.head.val == val:
            self.head = self.head.next
            self.size -= 1
            return True
        
        current = self.head
        while current.next:
            if current.next.val == val:
                current.next = current.next.next
                self.size -= 1
                return True
            current = current.next
        
        return False
    
    def search(self, val):
        """Search for value - O(n)"""
        current = self.head
        index = 0
        while current:
            if current.val == val:
                return index
            current = current.next
            index += 1
        return -1
    
    def reverse(self):
        """Reverse linked list - O(n)"""
        prev = None
        current = self.head
        
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        
        self.head = prev
    
    def find_middle(self):
        """Find middle element using two pointers - O(n)"""
        slow = fast = self.head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        return slow.val if slow else None
    
    def has_cycle(self):
        """Detect cycle using Floyd's algorithm - O(n)"""
        slow = fast = self.head
        
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        
        return False
    
    def display(self):
        """Display all elements"""
        result = []
        current = self.head
        while current:
            result.append(current.val)
            current = current.next
        return result


class DoublyListNode:
    """Node class for doubly linked list"""
    def __init__(self, val=0, prev=None, next=None):
        self.val = val
        self.prev = prev
        self.next = next


class DoublyLinkedList:
    """Doubly Linked List Implementation"""
    
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0
    
    def append(self, val):
        """Add node at the end - O(1)"""
        new_node = DoublyListNode(val)
        
        if not self.head:
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        
        self.size += 1
    
    def prepend(self, val):
        """Add node at the beginning - O(1)"""
        new_node = DoublyListNode(val)
        
        if not self.head:
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        
        self.size += 1
    
    def delete(self, val):
        """Delete first occurrence - O(n)"""
        current = self.head
        
        while current:
            if current.val == val:
                if current.prev:
                    current.prev.next = current.next
                else:
                    self.head = current.next
                
                if current.next:
                    current.next.prev = current.prev
                else:
                    self.tail = current.prev
                
                self.size -= 1
                return True
            
            current = current.next
        
        return False
    
    def display_forward(self):
        """Display from head to tail"""
        result = []
        current = self.head
        while current:
            result.append(current.val)
            current = current.next
        return result
    
    def display_backward(self):
        """Display from tail to head"""
        result = []
        current = self.tail
        while current:
            result.append(current.val)
            current = current.prev
        return result


# Example usage
if __name__ == "__main__":
    # Singly Linked List
    print("--- Singly Linked List ---")
    sll = SinglyLinkedList()
    sll.append(1)
    sll.append(2)
    sll.append(3)
    sll.prepend(0)
    print("List:", sll.display())
    print("Middle:", sll.find_middle())
    
    sll.reverse()
    print("Reversed:", sll.display())
    
    print("Search 2:", sll.search(2))
    sll.delete(2)
    print("After delete 2:", sll.display())
    
    # Doubly Linked List
    print("\n--- Doubly Linked List ---")
    dll = DoublyLinkedList()
    dll.append(1)
    dll.append(2)
    dll.append(3)
    dll.prepend(0)
    print("Forward:", dll.display_forward())
    print("Backward:", dll.display_backward())
    
    dll.delete(2)
    print("After delete 2 - Forward:", dll.display_forward())

